src/agent/solana/oracle.rs

```diff
@@ -43,6 +49,66 @@ use {
     },
 };
 
+/// This shim is used to abstract over SolanaPriceAccount and PythnetPriceAccount so we
+/// can iterate over either of these. The API is intended to force users to be aware of
+/// the account type they have, and so doesn't provide this abstraction (a good thing)
+/// and the agent should implement this in a better way.
+///
+/// For now, to implement the abstraction in the smallest way possible we use a shim
+/// type that uses the size of the accounts to determine the underlying representation
+/// and construct the right one regardless of which network we read. This will only work
+/// as long as we don't care about any extended fields.
+///
+/// TODO: Refactor the agent's network handling code.
+#[derive(Copy, Clone, Debug)]
+pub struct PriceEntry {
+    // We intentionally act as if we have a truncated account where the underlying memory is unavailable.
+    account: GenericPriceAccount<0, ()>,
+    comp:    [PriceComp; 64],
+}
+
+impl From<SolanaPriceAccount> for PriceEntry {
+    fn from(other: SolanaPriceAccount) -> PriceEntry {
+        unsafe {
+            // NOTE: We know the size is 32 because It's a Solana account. This is for tests only.
+            let comp_mem = std::slice::from_raw_parts(other.comp.as_ptr() as *const PriceComp, 32);
+            let account =
+                *(&other as *const SolanaPriceAccount as *const GenericPriceAccount<0, ()>);
+            let mut comp = [PriceComp::default(); 64];
+            comp.copy_from_slice(comp_mem);
+            PriceEntry { account, comp }
+        }
+    }
+}
+
+impl PriceEntry {
+    /// Construct the right underlying GenericPriceAccount based on the account size.
+    fn load_from_account(acc: &[u8]) -> Option<Self> {
+        unsafe {
+            let size = match acc.len() {
+                n if n == std::mem::size_of::<SolanaPriceAccount>() => 32,
+                n if n == std::mem::size_of::<PythnetPriceAccount>() => 64,
+                _ => return None,
+            };
```

Comment:
Do you really need unsafe? Why don't you conditionally deserialize as SolanaPriceAccount or PythnetPriceAccount.
Then you use the result SolanaPriceAccount or PythnetPriceAccount to build a PriceEntry.

Reply:
Can you expand on how to do that parsing? You might be right and I missed it but as far as I can tell these types are used mostly with zero copy casts and we haven't properly defined a parser/wire format to do this. In order to construct the underlying array I'm inferring from size, and the slice constructor is inherently unsafe which is where unsafe comes from.

Definitely would prefer a solution without unsafe, can you expand a bit more?

Reply:
Bytemuck try_from_bytes it like here : 
https://github.com/pyth-network/pyth-client/blob/f3b78939d768c525d29ed5fcebd767dc13e2fce8/program/rust/src/deserialize.rs#L33

Reply:
It's fine copying the data, you're copying it anyway